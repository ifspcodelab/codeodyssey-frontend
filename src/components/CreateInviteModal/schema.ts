import Yup from "../../config/yup.ts";

export const schema = Yup.object({
    endDate: Yup.date().required(),
}).required()
