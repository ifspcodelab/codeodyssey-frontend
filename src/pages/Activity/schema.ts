import Yup from "../../config/yup.ts";

export const schema = Yup.object({
    resolutionFile: Yup.mixed().required('File is required'),
}).required()
