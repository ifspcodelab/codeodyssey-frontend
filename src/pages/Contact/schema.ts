import Yup from "../../config/yup.ts";

export const schema = Yup.object({
    name: Yup.string().required().min(5).max(50),
    email: Yup.string().required().email(),
}).required()
