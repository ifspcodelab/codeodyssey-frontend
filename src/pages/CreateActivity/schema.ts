import Yup from "../../config/yup.ts";
import i18n from "../../locales/i18n";

export const schema = Yup.object({
    title: Yup.string().required().min(5).max(50),
    description: Yup.string().required().min(5),
    extension: Yup.string().required(),
    startDate: Yup.date().required(),
    endDate: Yup.date().required().min(Yup.ref('startDate')),
    initialFile: Yup.mixed().required(i18n.t('createactivity.form.validation.initialfile')),
    testFile: Yup.mixed().required(i18n.t('createactivity.form.validation.testfile')),
    solutionFile: Yup.mixed().required(i18n.t('createactivity.form.validation.solutionfile')),
}).required()
