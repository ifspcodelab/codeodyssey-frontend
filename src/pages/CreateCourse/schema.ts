import Yup from "../../config/yup.ts";
import i18n from "../../locales/i18n";


export const schema = Yup.object({
    name: Yup.string().required().min(1).max(255),
    slug: Yup.string().required().min(1).max(255).matches(
        /^[a-zA-Z0-9-]+$/,  i18n.t('createcourse.form.validation.special')),
    startDate: Yup.date().required(),
    endDate: Yup.date().required().min(Yup.ref('startDate'), i18n.t('createcourse.form.validation.endDate'))
}).required()
