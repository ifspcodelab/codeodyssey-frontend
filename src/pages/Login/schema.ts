import i18n from "../../locales/i18n";
import * as yup from "yup"

export const schema = yup
    .object({
        email: yup.string().required().email().max(350),
        password: yup
            .string()
            .required()
            .min(8)
            .max(64)
            .matches(/\d+/, i18n.t("login.form.validation.password.number"))
            .matches(/[a-z]+/, i18n.t("login.form.validation.password.lowercase"))
            .matches(/[A-Z]+/, i18n.t("login.form.validation.password.uppercase"))
            .matches(/[\W_]+/, i18n.t("login.form.validation.password.special")),
    })
    .required();