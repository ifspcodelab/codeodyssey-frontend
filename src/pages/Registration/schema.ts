import i18n from "../../locales/i18n";
import * as yup from "yup"

export const schema = yup.object({
    name: yup.string().required().min(5).max(100),
    email: yup.string().required().email().max(350),
    password: yup.string().required().min(8).max(64)
        .matches(/\d+/, i18n.t('registration.form.validation.password.number'))
        .matches(/[a-z]+/, i18n.t('registration.form.validation.password.lowercase'))
        .matches(/[A-Z]+/, i18n.t('registration.form.validation.password.uppercase'))
        .matches(/[\W_]+/, i18n.t('registration.form.validation.password.special')),
    terms: yup.boolean().oneOf([true], i18n.t('registration.form.validation.termsCheckbox'))
}).required()