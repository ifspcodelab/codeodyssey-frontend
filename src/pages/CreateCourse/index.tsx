import PageHeader from "../../components/PageHeader";
import { useTranslation } from "react-i18next";
import { Button, Container, Grid, TextField } from "@mui/material";
import { yupResolver } from "@hookform/resolvers/yup";
import { schema } from "./schema.ts";
import { SubmitHandler, useForm, Controller } from "react-hook-form";
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AxiosError } from "axios";
import dayjs from 'dayjs';
import { useNavigate } from "react-router-dom"
import { CreateCourseResponse } from "../../core/models/CreateCourseResponse";
import { CustomDate } from "../../core/models/CustomDate";
import ErrorSnackBar from "../../core/components/error-snack-bar/ErrorSnackBar.tsx";
import { AuthConsumer } from "../../core/auth/AuthContext.tsx";
import { JwtService } from "../../core/auth/JwtService.ts";
import 'dayjs/locale/pt-br'
import 'dayjs/locale/en'
import i18n from '../../locales/i18n.ts'
import { PageBaseLayout } from "../../core/layout/PageBaseLayout.tsx";
import { CoursesService } from "../../core/services/api/courses/CoursesService.ts";
import { useErrorHandler } from "../../core/hooks/useErrorHandler.ts";

function CreateCourse() {
  const { t } = useTranslation();
  const { handleError, openError, errorType, handleCloseError } = useErrorHandler();
  const authConsumer = AuthConsumer();
  const onSubmit: SubmitHandler<CreateCourseResponse> = (data) => submitCreateCourse(data)
  const { register, handleSubmit, watch, control, formState: { errors } } = useForm({ resolver: yupResolver(schema) })
  const navigate = useNavigate()
  const PROFESSOR_ID: string = authConsumer.id;
  const rawAccessToken = new JwtService().getRawAccessToken() as string;

  const submitCreateCourse = async (data: CreateCourseResponse) => {
    await CoursesService.create(data.name, data.slug, data.startDate.toISOString(), data.endDate.toISOString(), PROFESSOR_ID, rawAccessToken)
      .then(() => {
        navigate('/courses?success=true')
      }).catch((error: AxiosError<ProblemDetail>) => {
        handleError(error)
      })
  }

  const convertedDate: CustomDate = dayjs(new Date()) as unknown as CustomDate;

  return (
    <>
      <PageBaseLayout title={t('createcourse.title')}>
      </PageBaseLayout>
      <Container maxWidth="md">
        <PageHeader title={t('createcourse.title')} text={t('createcourse.text')} />
        <form onSubmit={handleSubmit(onSubmit)}>
          <Grid container spacing={1} rowSpacing={2}>
            <Grid item xs={12}>
              <TextField
                sx={{ width: "100%" }}
                {...register("name")}
                label={t("createcourse.form.name")}
                variant="outlined"
                error={!!errors.name}
                helperText={errors.name && <span>{errors.name.message}</span>}
                inputProps={{ "data-testid": "nameField" }}
              />
            </Grid>

            <Grid item xs={12}>
              <TextField
                sx={{ width: "100%" }}
                {...register("slug")}
                label={t("createcourse.form.slug")}
                variant="outlined"
                error={!!errors.slug}
                helperText={errors.slug && <span>{errors.slug.message}</span>}
                inputProps={{ "data-testid": "slugField" }}
              />
            </Grid>

            <Grid item xs={12} textAlign="right" display="flex" alignItems="spaceBetween">
              <Grid item xs={4}>
                <LocalizationProvider dateAdapter={AdapterDayjs} adapterLocale={i18n.language == "pt" ? "pt-br" : "en"} >
                  <Controller
                    name={"startDate"}
                    control={control}
                    defaultValue={convertedDate}

                    render={({ field: { ref, onChange, value, ...field } }) => (

                      <DatePicker
                        {...field}
                        inputRef={ref}
                        label={t("createcourse.form.startDate")} disablePast value={value ?? " "}
                        onChange={onChange as never}
                        slotProps={{
                          textField: {
                            helperText: errors.startDate && <span>{errors.startDate.message}</span>
                          },
                        }}
                      />
                    )}

                  />
                </LocalizationProvider>
              </Grid>

              <Grid >
                <LocalizationProvider dateAdapter={AdapterDayjs} adapterLocale={i18n.language == "pt" ? "pt-br" : "en"} >
                  <Controller
                    name={"endDate"}
                    control={control}
                    defaultValue={undefined}
                    render={({ field: { ref, onChange, value, ...field } }) => (

                      <DatePicker
                        {...field}
                        inputRef={ref}
                        minDate={watch().startDate}
                        label={t("createcourse.form.endDate")} data-testid="endDateField"
                        value={value ? value : null}
                        onChange={onChange as never}
                        slotProps={{
                          textField: {
                            helperText: errors.endDate && <span>{errors.endDate.message}</span>
                          },
                        }}
                      />
                    )}
                  />
                </LocalizationProvider>
              </Grid>
            </Grid>

            <Grid item xs={12} textAlign="right">
              <Button data-testid="submitButton" variant="outlined" type="submit">{t('createcourse.form.submit')}</Button>
            </Grid>
          </Grid>
        </form>

        <ErrorSnackBar open={openError} handleClose={handleCloseError} errorType={errorType} />
      </Container>
    </>
  )
}

export default CreateCourse