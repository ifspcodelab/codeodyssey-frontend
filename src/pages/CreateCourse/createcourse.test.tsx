import "@testing-library/jest-dom";
import CreateCourse from "./index"
import { render, fireEvent  } from "@testing-library/react";
import {describe, test, vi} from "vitest";
import {schema} from "../CreateCourse/schema.ts"
import {BrowserRouter} from "react-router-dom";

const cancelButtonId = "cancelButton"
const submitButtonId = "submitButton"
const nameFieldId = "nameField"
const slugFieldId = "slugField"
const startDateLabelName = "Start Date"
const endDateLabelName = "End Date"


describe("Create Course Form", () => {
  
  test("Should be able to see the title on the screen", () => {
    const {getByText} = render(
      <BrowserRouter>
        <CreateCourse/>
      </BrowserRouter>
    )

    expect(getByText("Create Course")).toBeInTheDocument();
  })

  test("Should be able to render the cancel button", () => {
    const {getByTestId} = render(
      <BrowserRouter>
        <CreateCourse/>
      </BrowserRouter>
    )

    expect(getByTestId(cancelButtonId)).toBeInTheDocument();
  })

  test("Should be able to render the submit button", () => {
    const {getByTestId} = render(
      <BrowserRouter>
        <CreateCourse/>
      </BrowserRouter>
    )

    expect(getByTestId(submitButtonId)).toBeInTheDocument();
  })

  test("Should name be validated", async () => {
    await expect(schema.validateAt('name', {name: ""})).rejects.toMatch(/This field is required./)
    await expect(schema.validateAt('name', {name: "react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react react "})).rejects.toMatch(/This field should be smaller than 255./)
    await expect(schema.validateAt('name', {name: "valid name"})).resolves.toBeTruthy()
  })

  test("Should slug be validated", async () => {
    await expect(schema.validateAt('slug', {slug: ""})).rejects.toMatch(/This field is required./)
    await expect(schema.validateAt('slug', {slug: "react native"})).rejects.toMatch(/This field cannot contain white space and special character/)
    await expect(schema.validateAt('slug', {slug: "react#3"})).rejects.toMatch(/This field cannot contain white space and special character/)
    await expect(schema.validateAt('slug', {slug: "validslug"})).resolves.toBeTruthy()
  })

  test("Should endDate be validated", async () => {
    await expect(schema.validateAt('endDate', {endDate: null})).rejects.toMatch(/This field is required./)
  })


  test("Should be able to render the form fields", () => {
    const {getByTestId, getByLabelText} = render(
      <BrowserRouter>
        <CreateCourse/>
      </BrowserRouter>
    )

    expect(getByTestId(nameFieldId)).toBeInTheDocument();
    expect(getByTestId(slugFieldId)).toBeInTheDocument();
    expect(getByLabelText(startDateLabelName)).toBeInTheDocument();
    expect(getByLabelText(endDateLabelName)).toBeInTheDocument();
  })

  test("Should be able to send create course form", () => {
    const { getByTestId } = render(
        <BrowserRouter>
            <CreateCourse/>
        </BrowserRouter>
    );

    fireEvent.click(getByTestId("submitButton"));
})

test("Should send request with data after form submission", async () => {
  const { getByTestId, getByLabelText } = render(
      <BrowserRouter>
          <CreateCourse />
      </BrowserRouter>
  );

  const nameInput = getByTestId("nameField")
  const slugInput = getByTestId('slugField')
  const startDate = getByLabelText(startDateLabelName)
  const endDate = getByLabelText(endDateLabelName)
  const submitButton =getByTestId('submitButton');

  fireEvent.change(nameInput, { target: { value: 'Java Spring' } });
  fireEvent.change(slugInput, { target: { value: 'java' } });
  fireEvent.change(startDate, { target: { value: '2023-12-01' } });
  fireEvent.change(endDate, { target: { value: '2024-01-01' } });

  fireEvent.click(submitButton);

  const result = await vi.fn();

  expect(result).toMatchSnapshot();
})

test("Should send to my courses page after clicking the submit button", async () => {
  const { getByTestId } = render(
      <BrowserRouter>
          <CreateCourse />
      </BrowserRouter>
  );

  const submitButton = getByTestId("submitButton")

  fireEvent.click(submitButton)

  const result = await vi.fn();

  expect(result).toMatchSnapshot("My Courses Page");
})
})