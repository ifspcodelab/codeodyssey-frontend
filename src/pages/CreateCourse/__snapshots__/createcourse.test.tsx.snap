// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Create Course Form > Should send to my courses page after clicking the submit button > My Courses Page 1`] = `[MockFunction spy]`;
